### PR TITLE
[FVM] Crypto Cadence updates 

### PR DESCRIPTION
**Pull request by janezpodhostnik**
References: https://github.com/dapperlabs/flow-go/issues/5543, https://github.com/dapperlabs/flow-go/issues/5544, https://github.com/dapperlabs/flow-go/issues/5541

Updated FVM part for the new Cadence crypto API

- support BLS and signature verification
- support KMAC and hashing with/without a tag
- update the public key decode interface (verify public key is now implemented, and doesn't always return false)

### CODE REVIEW

File: crypto/types.go [hunk 1]
```diff
@@ -124,6 +123,6 @@ func (e InvalidInputsError) Error() string {
 
 // IsInvalidInputsError checks if the input error is of a InvalidInputsError type
 func IsInvalidInputsError(err error) bool {
-	var errInvalidInputsError InvalidInputsError
-	return errors.As(err, &errInvalidInputsError)
+	_, ok := err.(*InvalidInputsError)
```
Comment (ramtinms) [inline crypto/types.go:126]: This prevents wrapped errors. For example a common case is `InvalidInputsError` wrapped into another higher level error level, and still in that case we want `IsInvalidInputsError` to return true. We mostly use Is method on higher level errors for this goal.

File: fvm/crypto/crypto.go [hunk 2]
```diff
@@ -174,46 +184,76 @@ func VerifySignatureFromRuntime(
...
+	message []byte,
+	publicKey crypto.PublicKey,
+	hashAlgo hash.HashingAlgorithm,
+) (bool, error) {
+
+	var hasher hash.Hasher
+
+	switch hashAlgo {
+	case hash.SHA2_256:
+		fallthrough
```
Comment (ramtinms) [inline fvm/crypto/crypto.go:241]: +1

File: fvm/crypto/crypto.go [hunk 3]
```diff
@@ -118,48 +94,82 @@ func CryptoToRuntimeHashingAlgorithm(h hash.HashingAlgorithm) runtime.HashAlgori
...
+			tag = flow.UserTagString
+		}
+		// tag compatibility
+		if !tagECDSACheck(tag) {
+			return false, errors.NewValueErrorf("tag %s is not supported", tag)
+		}
+	}
+
+	// check BLS compatibilites
+	if sigAlgo == crypto.BLSBLS12381 && hashAlgo != hash.KMAC128 {
```
Comment (turbolent) [inline fvm/crypto/crypto.go:161]: Maybe use else-if

File: fvm/crypto/crypto.go [hunk 4]
```diff
@@ -174,46 +184,76 @@ func VerifySignatureFromRuntime(
...
+	publicKey crypto.PublicKey,
+	hashAlgo hash.HashingAlgorithm,
+) (bool, error) {
+
+	var hasher hash.Hasher
+
+	switch hashAlgo {
+	case hash.SHA2_256:
+		fallthrough
+	case hash.SHA3_256:
```
Comment (turbolent) [inline fvm/crypto/crypto.go:242]:
```suggestion
	case hash.SHA2_256, hash.SHA3_256:
```

**Review by ramtinms — APPROVED** [review]
Except the minor comment about AsError the rest looks good to me

**Review by turbolent — APPROVED** [review]
Nice!

File: crypto/types.go (hunk 1 above)
Comment (tarakby) [reply crypto/types.go:126]: You're right. I changed this for 2 reasons though:
 - there are no wrapped errors in this case.
 - `errors.As` as implemented previously wasn't giving the correct behaviour, I'm happy to use it back if there is a fix.

**Pull request merged**